### ISSUE TITLE
Wrong check in optimize_for_inference

### ISSUE THREAD

**Issue by spiegelball** [pytorch/pytorch#68582]
## 🐛 Bug
https://github.com/pytorch/pytorch/blob/dbbb02474b9ad8044d061caee9fb0916da93101b/torch/csrc/jit/api/module.cpp#L471-L477

The check in line 475 prevents an evaluated model to be optimized. Is this correct?

**Comment by Linux-cpp-lisp** [comment]
We have (I think) this issue as well:
```c++
model.eval();
model = torch::jit::freeze(model);
```
gives
```
terminate called after throwing an instance of 'c10::Error'
  what():  Freezing is currently only implemented for modules in eval mode. Please call .eval() before freezing
```

**Comment by Linux-cpp-lisp** [comment]
OK, this appears to be a version thing— in 1.10 the check is correct (`!model.is_training()`), but in 1.9 and I presume earlier, it is just `model.is_training()`, giving the issue.

**Comment by spiegelball** [comment]
Do you have a link to the code? I just checked on the 1.10 tag and its still not correct.

https://github.com/pytorch/pytorch/blob/36449ea93134574c2a22b87baad3de0bf8d64d42/torch/csrc/jit/api/module.cpp#L493

**Comment by Linux-cpp-lisp** [comment]
that's the link I'm looking at... I think it's right in 1.10, since it checks first that asking the module `is_training()` is sane (otherwise the `!hasattr` will short circut), and then checks that it is _not_ in training mode?

**Comment by eellison** [comment]
Sounds like this is fixed ? If not could you please give a full repro on the current master. Sorry for the earlier trouble.
